server/src/main/java/org/opensearch/action/bulk/TransportBulkAction.java

```diff
@@ -756,6 +762,44 @@ public void onTimeout(TimeValue timeout) {
             });
         }
 
+        private boolean addFailureIfIndexAppendOnlyAndOpsDeleteOrUpdate(
+            DocWriteRequest<?> request,
+            int idx,
+            final ConcreteIndices concreteIndices,
+            Metadata metadata
+        ) {
+            Index concreteIndex = concreteIndices.getConcreteIndex(request.index());
+            if (concreteIndex == null) {
+                try {
+                    concreteIndex = concreteIndices.resolveIfAbsent(request);
+                } catch (IndexClosedException | IndexNotFoundException | IllegalArgumentException ex) {
+                    addFailure(request, idx, ex);
+                    return true;
+                }
+            }
+
+            final IndexMetadata indexMetadata = metadata.index(concreteIndex);
+            if (indexMetadata.isAppendOnlyIndex() == true) {
+                if ((request.opType() == DocWriteRequest.OpType.UPDATE || request.opType() == DocWriteRequest.OpType.DELETE)) {
+                    ValidationException exception = new ValidationException();
+                    exception.addValidationError(
+                        "Operation [" + request.opType() + "] is not allowed for append only index " + request.index()
+                    );
+                    addFailure(request, idx, exception);
+                    return true;
+                } else if (request.id() != null && request.opType() == DocWriteRequest.OpType.INDEX) {
+                    ValidationException exception = new ValidationException();
+                    exception.addValidationError(
+                        "Operation [" + request.opType() + "] is not allowed with a custom doc id " + request.id()
```

Comment:
Can we add to the message that the reason validation is failing because append only setting is set to true?

Reply:
Ack.